src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,428 @@
+                            "trace_parent_id": span.get_span_context().span_id,
+                            "is_flow_log": False,
+                        }
+                    )
+
+                # Write the Tool Kernel to the Span on HL_FILE_OT_KEY
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HL_FILE_OT_KEY,
+                    value={
+                        "path": path if path else func.__name__,
+                        "tool": tool_kernel,
+                    },
+                )
+
+                # Call the decorated function
+                output = func(*args, **kwargs)
+
+                # All children Spans have been created when the decorated function returns
+                # Remove the Trace metadata from the context so the siblings can have
+                # their children linked properly
+                if trace_metadata:
+                    pop_trace_context()
+
+                # Populate known Tool Log attributes
+                tool_log = {
+                    "inputs": args_to_inputs(func, args, kwargs),
+                }
+                if output:
+                    tool_log["output"] = output
+
+                # Write the Tool Log to the Span on HL_LOG_OT_KEY
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HL_LOG_OT_KEY,
+                    value=tool_log,
+                )
+
+                # Return the output of the decorated function
+                return output
+
+        return wrapper
+
+    return decorator
+
+
+def _build_tool_kernel(
+    func: Callable,
+    attributes: Optional[dict[str, Optional[Any]]],
+    setup_values: Optional[dict[str, Optional[Any]]],
+    strict: bool,
+) -> ToolKernelRequestParams:
+    """Build ToolKernelRequest object from decorated function."""
+    return ToolKernelRequestParams(
+        source_code=textwrap.dedent(
+            # Remove the tool decorator from source code
+            inspect.getsource(func).split("\n", maxsplit=1)[1]
+        ),
+        # Note: OTel complains about falsy values in attributes, so we use OT_EMPTY_ATTRIBUTE
+        attributes=attributes or HL_OT_EMPTY_VALUE,  # type: ignore
+        setup_values=setup_values or HL_OT_EMPTY_VALUE,  # type: ignore
+        function=_build_function_property(
+            func=func,
+            strict=strict,
+        ),
+    )
+
+
+def _build_function_property(func: Callable, strict: bool) -> ToolFunctionParams:
+    """Build `function` property inside ToolKernelRequest."""
+    tool_name = func.__name__
+    description = func.__doc__
+    if description is None:
+        description = ""
+    return ToolFunctionParams(
+        name=tool_name,
+        description=description,
+        parameters=_build_function_parameters_property(func),  # type: ignore
+        strict=strict,
+    )
+
+
+class _JSONSchemaFunctionParameters(TypedDict):
```

Comment:
I don't think this is a complete spec.

My understanding is that JSONSchema can have arrays as properties

Reply:
ack, will fix

Reply:
Was this resolved?

Reply:
we support extracting function arguments that are lists